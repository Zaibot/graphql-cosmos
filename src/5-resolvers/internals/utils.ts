import { FeedResponse } from '@azure/cosmos'
import { GraphQLCosmosPageOutput } from '../../4-resolver-builder/3-typedefs-transformer'
import { Lazy } from '../../typescript'
import { SourceDescriptor } from '../x-descriptors'

export const emptyPageResponse: GraphQLCosmosPageOutput<any> = {
  cursor: null,
  nextCursor: null,
  page: [],
  total: 0,
}

export function graphqlCosmosPageResponse<T extends { id: string }>(
  cursor: string | null,
  query: Lazy<Promise<FeedResponse<T>>>,
  count: Lazy<Promise<FeedResponse<number>>>,
  transformer: (item: T) => SourceDescriptor.Embedded<T>
): GraphQLCosmosPageOutput<T> {
  return {
    cursor: cursor,
    nextCursor: () => query().then((x) => x.continuationToken),
    page: () => query().then((x) => (x.resources ?? []).map(transformer)),
    total: () => count().then((x) => x.resources[0]),
  }
}
