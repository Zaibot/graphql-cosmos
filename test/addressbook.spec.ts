import gql from 'graphql-tag'
import { printSchemaWithDirectives } from 'graphql-tools'
import { createUnitTestContext } from './utils'

describe(`Address Book`, () => {
  const dummyTypeDefs = gql`
    type Query {
      person: Person @cosmos
      contact: Contact @cosmos
      people: [Person!]! @cosmos
    }

    type Person @cosmos(database: "AddressBook", container: "Persons") {
      id: ID @where(op: "eq")
      name: String! @where(op: "contains_lowercase")
      contacts: [Contact!]! @cosmos(theirs: "personId", pagination: "off")
      perferredContact: Contact @cosmos(ours: "preferredContactId")
    }

    type Contact @cosmos(database: "AddressBook", container: "Contacts") {
      id: ID @where(op: "eq")
      person: Person! @cosmos(ours: "personId")
      preferredBy: [Person!]! @cosmos(theirs: "preferredContactId", pagination: "off")
      phonenumber: String!
    }
  `

  const responses = {
    Persons: {
      'SELECT c.id FROM c ORDER BY c.id': [{ id: `alice` }, { id: `bob` }],
      'SELECT c.id, c.name FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=alice,bob)': [
        { id: `alice`, name: `Alice` },
        { id: `bob`, name: `Bob` },
      ],
    },
    Contacts: {
      'SELECT c.id FROM c WHERE c.personId = @p2 ORDER BY c.id (@p2=alice)': [{ id: `alice` }],
      'SELECT c.id FROM c WHERE c.personId = @p2 ORDER BY c.id (@p2=bob)': [{ id: `bob` }],
      'SELECT c.id, c.phonenumber FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=alice,bob)': [
        { id: `alice`, phonenumber: `555-111` },
        { id: `bob`, phonenumber: `555-222` },
      ],
    },
  }

  const uc = createUnitTestContext(dummyTypeDefs, responses)

  it(`expects schema to remain the same`, () => {
    const output = printSchemaWithDirectives(uc.schema)
    expect(output).toMatchSnapshot()
  })

  it(`expects meta schema to remain the same`, () => {
    const output = uc.metaSchema
    expect(output).toMatchSnapshot()
  })

  it(`should be retrieve all items`, async () => {
    const result = await uc.execute(`query { people { page { id name contacts { phonenumber } } } } `)

    expect(result).toEqual({
      data: {
        people: {
          page: [
            {
              contacts: [{ phonenumber: '555-111' }],
              id: 'alice',
              name: 'Alice',
            },
            {
              contacts: [{ phonenumber: '555-222' }],
              id: 'bob',
              name: 'Bob',
            },
          ],
        },
      },
    })
  })
})
