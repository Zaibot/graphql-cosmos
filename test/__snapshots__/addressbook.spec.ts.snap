// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address Book expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": "Contacts",
      "cosmos": true,
      "database": "AddressBook",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
          ],
          "whereOurs": null,
        },
        Object {
          "container": "Persons",
          "cosmos": true,
          "database": "AddressBook",
          "fieldname": "person",
          "ours": "personId",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Person",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "phonenumber",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Persons",
          "cosmos": true,
          "database": "AddressBook",
          "fieldname": "preferredBy",
          "ours": null,
          "pagination": false,
          "returnMany": true,
          "returnTypename": "Person",
          "sortOurs": null,
          "sortable": false,
          "theirs": "preferredContactId",
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": true,
      "sortable": false,
      "typename": "Contact",
    },
    Object {
      "container": "Persons",
      "cosmos": true,
      "database": "AddressBook",
      "external": false,
      "fields": Array [
        Object {
          "container": "Contacts",
          "cosmos": true,
          "database": "AddressBook",
          "fieldname": "contacts",
          "ours": null,
          "pagination": false,
          "returnMany": true,
          "returnTypename": "Contact",
          "sortOurs": null,
          "sortable": false,
          "theirs": "personId",
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
          ],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "name",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "containslower",
          ],
          "whereOurs": null,
        },
        Object {
          "container": "Contacts",
          "cosmos": true,
          "database": "AddressBook",
          "fieldname": "perferredContact",
          "ours": "preferredContactId",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Contact",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": true,
      "sortable": false,
      "typename": "Person",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": "Contacts",
          "cosmos": true,
          "database": "AddressBook",
          "fieldname": "contact",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Contact",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Persons",
          "cosmos": true,
          "database": "AddressBook",
          "fieldname": "people",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Person",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Persons",
          "cosmos": true,
          "database": "AddressBook",
          "fieldname": "person",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Person",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Query",
    },
  ],
}
`;

exports[`Address Book expects schema to remain the same 1`] = `
"schema {
  query: Query
}
type Query {
  person(where: PersonWhere): Person
  contact(where: ContactWhere): Contact
  people(where: PersonWhere, cursor: String, limit: Int): PersonPage!
}
type Person {
  id: ID
  name: String!
  contacts(where: ContactWhere, limit: Int): [Contact!]!
  perferredContact: Contact
}
type Contact {
  id: ID
  person: Person!
  preferredBy(where: PersonWhere, limit: Int): [Person!]!
  phonenumber: String!
}
type PersonPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Person]!
}
input ContactWhere {
  and: [ContactWhere!]
  or: [ContactWhere!]
  id_eq: ID
}
input PersonWhere {
  and: [PersonWhere!]
  or: [PersonWhere!]
  id_eq: ID
  name_containslower: String
}"
`;
